ambry-cloud/src/main/java/com.github.ambry.cloud/CloudMessageReadSet.java

```diff
@@ -0,0 +1,135 @@
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+package com.github.ambry.cloud;
+
+import com.github.ambry.commons.BlobId;
+import com.github.ambry.store.MessageReadSet;
+import com.github.ambry.store.StoreKey;
+import com.github.ambry.utils.ByteBufferOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.util.List;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+class BlobReadInfo {
+  private final CloudBlobMetadata blobMetadata;
+  private final BlobId blobId;
+  private ByteBuffer prefetchedBuffer;
+  private boolean isPrefetched;
+
+  public BlobReadInfo(CloudBlobMetadata blobMetadata, BlobId blobId) {
+    this.blobMetadata = blobMetadata;
+    this.blobId = blobId;
+    this.isPrefetched = false;
+  }
+
+  public void prefetchBlob(CloudDestination cloudDestination) throws CloudStorageException {
+    prefetchedBuffer = ByteBuffer.allocate((int) blobMetadata.getSize());
+    ByteBufferOutputStream outputStream = new ByteBufferOutputStream(prefetchedBuffer);
+    cloudDestination.downloadBlob(blobId, outputStream);
+    isPrefetched = true;
+  }
+
+  public ByteBuffer downloadBlob(CloudDestination cloudDestination) throws CloudStorageException {
+    ByteBuffer blobByteBuffer = ByteBuffer.allocate((int) blobMetadata.getSize());
+    ByteBufferOutputStream outputStream = new ByteBufferOutputStream(blobByteBuffer);
+    cloudDestination.downloadBlob(blobId, outputStream);
+    return blobByteBuffer;
+  }
+
+  public ByteBuffer getPrefetchedBuffer() {
+    return prefetchedBuffer;
+  }
+
+  public boolean isPrefetched() {
+    return isPrefetched;
+  }
+
+  public BlobId getBlobId() {
+    return blobId;
+  }
+
+  public long getBlobSize() {
+    return blobMetadata.getSize();
+  }
+}
+
+class CloudMessageReadSet implements MessageReadSet {
+
+  private final List<BlobReadInfo> blobReadInfoList;
+  private final CloudDestination cloudDestination;
+  private final Logger logger = LoggerFactory.getLogger(getClass());
+
+  CloudMessageReadSet(List<BlobReadInfo> blobReadInfoList, CloudDestination cloudDestination) {
+    this.blobReadInfoList = blobReadInfoList;
+    this.cloudDestination = cloudDestination;
+  }
+
+  @Override
+  public long writeTo(int index, WritableByteChannel channel, long relativeOffset, long maxSize) throws IOException {
+    validateIndex(index);
+    long written = 0;
+    BlobReadInfo blobReadInfo = blobReadInfoList.get(index);
+
+    ByteBuffer outputBuffer;
+    try {
+      if (blobReadInfo.isPrefetched()) {
+        outputBuffer = blobReadInfo.getPrefetchedBuffer();
+      } else {
+        outputBuffer = blobReadInfo.downloadBlob(cloudDestination);
```

Comment:
In the non-prefetch case, we shouldn't need to write the outputsteam to an in memory buffer that will be passed to channel.  It can be streamed instead.

Reply:
fixed.